src/effects/native/gatereffect.cpp

```diff
@@ -0,0 +1,236 @@
+    quantize->setId("quantize");
+    quantize->setName("Quantize");
+    quantize->setShortName("Quantize");
+    quantize->setDescription(
+        "Round the Time parameter to the nearest 1/4 beat.");
+    quantize->setControlHint(
+        EffectManifestParameter::ControlHint::TOGGLE_STEPPING);
+    quantize->setSemanticHint(EffectManifestParameter::SemanticHint::UNKNOWN);
+    quantize->setUnitsHint(EffectManifestParameter::UnitsHint::UNKNOWN);
+    quantize->setDefault(1);
+    quantize->setMinimum(0);
+    quantize->setMaximum(1);
+
+    EffectManifestParameter* triplet = manifest.addParameter();
+    triplet->setId("triplet");
+    triplet->setName("Triplets");
+    triplet->setDescription("When the Quantize parameter is enabled, divide "
+                            "rounded 1/4 beats of Time parameter by 3.");
+    triplet->setControlHint(
+        EffectManifestParameter::ControlHint::TOGGLE_STEPPING);
+    triplet->setSemanticHint(EffectManifestParameter::SemanticHint::UNKNOWN);
+    triplet->setUnitsHint(EffectManifestParameter::UnitsHint::UNKNOWN);
+    triplet->setDefault(0);
+    triplet->setMinimum(0);
+    triplet->setMaximum(1);
+
+    EffectManifestParameter *invert = manifest.addParameter();
+    invert->setId("invert");
+    invert->setName("Invert");
+    invert->setDescription("");
+    invert->setControlHint(
+        EffectManifestParameter::ControlHint::TOGGLE_STEPPING);
+    invert->setSemanticHint(EffectManifestParameter::SemanticHint::UNKNOWN);
+    invert->setUnitsHint(EffectManifestParameter::UnitsHint::UNKNOWN);
+    invert->setDefault(0);
+    invert->setMinimum(0);
+    invert->setMaximum(1);
+
+    return manifest;
+}
+
+GaterEffect::GaterEffect(EngineEffect* pEffect,
+                             const EffectManifest& manifest)
+        : m_pRateParameter(pEffect->getParameterById("rate")),
+          m_pShapeParameter(pEffect->getParameterById("shape")),
+          m_pQuantizeParameter(pEffect->getParameterById("quantize")),
+          m_pTripletParameter(pEffect->getParameterById("triplet")),
+          m_pInvertParameter(pEffect->getParameterById("invert")) 
+{
+    Q_UNUSED(manifest);
+}
+
+GaterEffect::~GaterEffect() {
+}
+
+void GaterEffect::processChannel(const ChannelHandle& handle,
+                                GaterGroupState* pState,
+                                const CSAMPLE* pInput, CSAMPLE* pOutput,
+                                const unsigned int numSamples,
+                                const unsigned int sampleRate,
+                                const EffectProcessor::EnableState enableState,
+                                const GroupFeatureState& groupFeatures) { 
+                                    
+    Q_UNUSED(handle);
+                                    
+    const double rate  = m_pRateParameter->value();
+    const double shape  = m_pShapeParameter->value();
+
+    // Used to divide a beat by a power of 2
+    int divider = pow(2, floor(3*rate));
```

Comment:
I don't understand this scaling. It feels random. Could you explain your reasoning for this?

Reply:
The aim here is to have a "round" division of beats, that is, an effect period of 1/8, 1/4, 1/2 or 1 beat (when quantization is on of course), as this keeps the effect aligned with beats.
This divider is bypassed when quantization is off